### ISSUE TITLE
incompatible with windows

### ISSUE THREAD

**Issue by 2bndy5** [adafruit/circuitpython-build-tools#48]
I've been looking into the error I get on windows pertaining to [line 200 of build.py](https://github.com/sommersoft/circuitpython-build-tools/blob/800ee1a567ee59b107a738cb953fa3dd0878b7ce/circuitpython_build_tools/build.py#L200). Everytime I run circuitpython-build-bundles I get a permission denied error in windows:
`PermissionError: [Errno 13] Permission denied: 'C:\\Users\\<my name>\\AppData\\Local\\Temp\\tmpzokqcaha'` I've been backtracking the creation of the temp file and found [this little tidbit in the tempfile docs](https://docs.python.org/3.7/library/tempfile.html#tempfile.NamedTemporaryFile): 

> Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms (it can be so used on Unix; it cannot on Windows NT or later)

so, it looks like you can't copy the file while it is still open (inside the with block) on windows... I'll fork the repo and test my idea for a solution and get back to you (unless someone beats me to it).

PS. Strangely, I cannot find the file it specifies in my appdata/local/temp folder, but I think its because tempfile.NamedTemporaryFile() deletes it when exiting the with block on the raised exception.

**Comment by tannewt** [comment]
Thanks for taking a look at this! We don't expect many folks to run it themselves. WSL might be the easiest way for windows users.

**Comment by 2bndy5** [comment]
Oh yeah! Good call, I'll have to check that out



Get BlueMail for Android

**Comment by 2bndy5** [comment]
### follow-up (just in case someone lands here from a google search)

WSL works, but you need to run the following (using Ubuntu WSL):

1. `sudo apt install make python3 python3-pip gettext`. 
    - `make` might come with Ubuntu WSL, but it doesn't hurt to ensure it is installed.
    - `gettext` is required so that `make` (called internally by this lib) can build mpy-cross. Otherwise, `make` outputs an error that reads `msgfmt Command not recognized` (or something like that pertaining to a `msgfmt` command).
2. `pip3 install circuitpython-build-tools`. (duh)
    - pip may complain that python's script folder is not in the PATH environment variable (see step 4)
3. `sudo update-alternatives --install /usr/bin/python python /usr/bin/python3 10`
    - this properly aliases python3 as python. Required because `circuitpython-buld-bundles` script calls python, not python3
4. `export PATH=$PATH:<scripts_folder>` where `<scripts_folder>` is the folder that pip identifies as not part of the PATH environment variable.
    - If pip doesn't complain about the scripts folder from step 2, then skip this.

**Comment by jepler** [comment]
If what circuitpython-build-tools needs is mpy-cross (and not other parts of circuitpython) maybe we can revise it to use the windows, linux, and mac static builds of mpy-cross that are created as build artifacts in CircuitPython, instead of fetching and building the source itself.

That doesn't fix the NamedTemporaryFile problem, of course.

**Comment by 2bndy5** [comment]
Fixing the NamedTemporaryFile problem wasn't hard. I just had to move a line out of the with block to let the file in use be closed before copying it, but I have since deleted my fork when this issue was closed.
Using static builds of mpy-cross would be appreciated as cloning CircuitPython repo comes out to an additional 200 MB or so (including submodules) just to build mpy-cross. It does seem to build mpy-cross for different releases of CircuitPython, so I don't know if that complicates things further.
